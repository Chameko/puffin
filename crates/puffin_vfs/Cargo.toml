[package]
name = "puffin_vfs"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
crossbeam-channel = "0.5.8"
fxhash = "0.2"
notify = "5.1.0"
relative-path = { version = "1.8", default-features = false }
thiserror = "1.0"
