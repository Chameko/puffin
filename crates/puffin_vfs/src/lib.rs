/// Contains the path interner
mod path_interner;
use path_interner::PathInterner;

/// Contains the moniter
mod monitor;
pub use monitor::Monitor;
pub use monitor::ForegroundMessage;
pub use monitor::MonitorMessage;

/// Contains the path types used in Puffin
pub mod path;
pub use path::AbsPath;
pub use path::AbsPathBuf;

use relative_path::RelativePath;
use thiserror::Error;

/// Identifier for a file in the virtual file system
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord)]
#[repr(transparent)]
pub struct FileID(pub u32);

/// Error type for the virtual file system
#[derive(Error, Debug)]
pub enum VFSError {
    #[error("expected an absolute path")]
    NotAbsolutePath,
}

/// The virtual file system
#[derive(Debug)]
pub struct VFS {
    /// The internal path interner
    interner: PathInterner,
}

impl VFS {
    /// Either adds a path and returns the [`FileID`] or returns the [`FileID`] if it already exists
    pub fn intern(&mut self, path: &AbsPath) -> FileID {
        self.interner.intern(path)
    }

    /// Get an absolute path from a file ID
    pub fn get_path(&self, id: FileID) -> Option<&AbsPath> {
        self.interner.get_path(id)
    }

    /// Intern a [RelativePath] that is relative to a supplied [AbsPath]
    pub fn intern_relative_path(&mut self, path: &RelativePath, relative_to: &AbsPath) -> FileID {
        self.interner.intern(&AbsPathBuf::try_from(path.to_logical_path(relative_to)).unwrap())
    }
}
