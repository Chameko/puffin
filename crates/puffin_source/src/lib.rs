pub mod module;
pub mod id;

use puffin_vfs::VFS;
pub use puffin_vfs::FileID;
use relative_path::RelativePathBuf;
use std::sync::Arc;
use crate::module::{ModuleTree, generate_module_tree};

/// Type use for marking slices of text in the source files
pub type TextSlice = std::ops::RangeInclusive<u32>;

/// The query group for interacting with the source files
#[salsa::query_group(SourceStorage)]
pub trait SourceDatabase {
    /// Set the source tree
    #[salsa::input]
    fn source_tree(&self) -> Arc<SourceTree>;

    /// Set the [VFS]
    #[salsa::input]
    fn vfs(&self) -> Arc<VFS>;

    #[salsa::invoke(generate_module_tree)]
    fn generate_module_tree(&self) -> Arc<ModuleTree>;

    /// Get the relative path of a file to the source tree
    fn relative_path(&self, file_id: FileID) -> Option<RelativePathBuf>;
}

/// Get the relative path of a file to the source tree
fn relative_path(db: &dyn SourceDatabase, file_id: FileID) -> Option<RelativePathBuf> {
    let tree = db.source_tree();
    let vfs = db.vfs();
    tree.get_relative_path(file_id, vfs)
}

/// The source tree represents a tree of sources in a Puffin project. Each file is relative to the source tree
#[derive(Debug, Clone)]
pub struct SourceTree {
    /// The base directory of the source tree. Everything else is relative to this
    dir: FileID,
    /// The sources in the source tree
    pub sources: Vec<Source>
}

impl SourceTree {
    /// Find a source in the source tree
    pub fn find_source(&self, file_id: FileID) -> Option<&Source> {
        for source in &self.sources {
            if source.file == file_id {
                return Some(source)
            }
        }
        None
    }

    /// Get the path of a source relative to this source tree
    pub fn get_relative_path(&self, file_id: FileID, vfs: Arc<VFS>) -> Option<RelativePathBuf> {
        let src_path = self.find_source(file_id).map(|s| vfs.get_path(s.file)).flatten();
        let path = vfs.get_path(self.dir);
        if let (Some(src_path), Some(path)) = (src_path, path) {
            if let Ok(path) = src_path.strip_prefix(path) {
                Some(RelativePathBuf::from(path.as_os_str().to_string_lossy().to_string()))
            } else {
                None
            }
        } else {
            None
        }
    }
}


/// A source in the source tree
#[derive(Debug, Clone, PartialEq)]
pub struct Source {
    // The file that the source belongs to
    pub file: FileID,
    // The text contained within the file
    pub text: String,
}

impl Source {
    /// Get the text of a source file
    pub fn get_text(&self) -> &str {
        self.text.as_str()
    }
}
