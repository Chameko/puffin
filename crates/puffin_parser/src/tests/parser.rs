use puffin_source::Source;
use puffin_vfs::FileID;
use crate::{TokenStream, lexer::Lexer, parser::Parser};
use super::output_cst;

/// Parses the inputed string and creates a snapshot
fn standard_test(src: &str) -> String {
    let lexer = Lexer::new(src);
    let source = Source{ file: FileID(0), text: src.to_string() };
    let parser = Parser::new(TokenStream::new(lexer.start_scan()), &source);
    let parse = parser.parse_test();
    if parse.errors.len() > 0 {
        for error in parse.errors {
            println!("{}", error.debug_display("test.pf", src.split_inclusive('\n').enumerate().map(|mut v| {
                v.0 += 1;
                v
            }).collect()));
            panic!("Errors detected when parsing");
        }
    }
    output_cst(&parse.green_node)
}

/// Same as standard_test except it uses the default parsing function
fn item_test(src: &str) -> String {
    let lexer = Lexer::new(src);
    let source = Source{ file: FileID(0), text: src.to_string() };
    let parser = Parser::new(TokenStream::new(lexer.start_scan()), &source);
    let parse = parser.parse();
    if parse.errors.len() > 0 {
        for error in parse.errors {
            println!("{}", error.debug_display("test.pf", src.split_inclusive('\n').enumerate().map(|mut v| {
                v.0 += 1;
                v
            }).collect()));
            panic!("Errors detected when parsing");
        }
    }
    output_cst(&parse.green_node)
}

#[test]
fn just_number() {
    insta::assert_snapshot!(standard_test("1"));
}

#[test]
fn simple_expr() {
    insta::assert_snapshot!(standard_test("1 + 2"));
}

#[test]
fn multiple_simple() {
    insta::assert_snapshot!(standard_test("1 + 2 + 3 + 4"));
}

#[test]
fn prefix_operation() {
    insta::assert_snapshot!(standard_test("-1 + 2"));
}

#[test]
fn multiple_prefix_operation() {
    insta::assert_snapshot!(standard_test("--1 + -2"));
}

#[test]
fn order_of_operations() {
    insta::assert_snapshot!(standard_test("1 + 2 * 5 - 3"));
}

#[test]
fn paren() {
    insta::assert_snapshot!(standard_test("1 + 2 * (3 - 2) + (1 * 2)"));
}

#[test]
fn paren_2() {
    insta::assert_snapshot!(standard_test("(1 + 2 + 3)"));
}

#[test]
fn paren_3() {
    insta::assert_snapshot!(standard_test("(1 + (1)) - (2 + 1 + 2)"));
}

#[test]
fn print_stmt() {
    insta::assert_snapshot!(standard_test("print 2 * (4 - 2)"));
}

#[test]
fn multiple_stmt() {
    insta::assert_snapshot!(standard_test("1 + 2\nprint 3\n2 * 4"));
}

#[test]
fn let_stmt() {
    insta::assert_snapshot!(standard_test("let banana = 1"));
}

#[test]
fn block_stmt() {
    insta::assert_snapshot!(standard_test("{ print 1 + 1 }"));
}

#[test]
fn block_stmt_2() {
    insta::assert_snapshot!(standard_test("{\nprint 1 + 1\n}"));
}

#[test]
fn block_stmt_3() {
    insta::assert_snapshot!(standard_test("{print 1 + 1\n}"));
}

#[test]
fn block_stmt_4() {
    insta::assert_snapshot!(standard_test("{\nprint 1 + 1}"));
}

#[test]
fn assign_stmt() {
    insta::assert_snapshot!(standard_test("a = 4 + 3"));
}

#[test]
fn let_stmt_concrete_type() {
    insta::assert_snapshot!(standard_test("let a: float = 2.0"));
}

#[test]
fn basic_fn() {
    insta::assert_snapshot!(item_test("fun hello() {}"));

}
#[test]
fn fn_with_param() {
    insta::assert_snapshot!(item_test("fun move(p1: Point, p2: Point) {}"));
}

#[test]
fn multiple_assign() {
    insta::assert_snapshot!(standard_test("a = b = c = 1"));
}
