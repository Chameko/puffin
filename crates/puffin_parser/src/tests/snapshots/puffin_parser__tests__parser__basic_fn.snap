---
source: crates/puffin_parser/src/tests/parser.rs
expression: "item_test(\"fun hello() {}\")"
---
SOURCE_FILE@0..=13
    FUNC_ITEM@0..=13
        fun |KW_FUN@0..2
          |WHITESPACE@3..3
        hello |IDENT@4..8
        FUNC_PARAM@9..=10
            ( |L_PAREN@9..9
            ) |R_PAREN@10..10
          |WHITESPACE@11..11
        BLOCK_STMT@12..=13
            { |L_BRACE@12..12
            } |R_BRACE@13..13

